### ISSUE TITLE
Thread-safety of logging?

### ISSUE THREAD

**Issue by maltezfaria** [maltezfaria/DataFlowTasks.jl#59]
Because of the recent thread migration system in Julia, I don't think the following lines in the logging mechanics are thread-safe:

https://github.com/maltezfaria/DataFlowTasks.jl/blob/9e4c09319611f0b5663c00c59f6b7332da16cb27/src/dataflowtask.jl#L52-L56

The reason is that when a `Task` actually pushes into the `tasklogs` vector, it may live on a different thread than its `tid` variable. As such, two tasks on two different threads, but with the same value of `tid` could push concurrently to the `Vector{TaskLog}` used for logging, creating a race condition. 

The current design has `taskslogs` as vector of vectors of `TasksLog`s. After searching the code, most of the use of `tasklogs` actually just flattens that structure using `Iterators.flatten` in order to loop over all `TaskLog`s, so I am assuming that the original design was so that different tasks could just push to the vector-of-vectors structure without using any locks. But as mentioned above, this may no longer be *safe*. An alternative is to replace the vector-of-vectors simply by a `Channel{TaskLog}`. This may induce some overhead when the logging is active, but I doubt it will be anything major (also performance is not of uttermost importance when logging IMO).

**Comment by maltezfaria** [comment]
While IMO the concern above is valid, I think the code block mentioned (while arguably brittle) is safe. The reason is that, under the assumption that no task switches occur in the constructor `TaskLog(...)` (e.g. it does not `wait` or perform an `IO`), the value of `tid` will indeed be the id of the thread that executes the `push!` in the last line. 

@ffevotte Based on the argument above, I am tempted to close this issue. Any thoughts?

**Comment by ffevotte** [comment]
I agree. I have been able to reproduce the `KeyError` issues locally, and to check that no task migration occurred.

Maybe we should simply put some a posteriori checks regarding task migrations, and display a warning when they occur?
Even something crude like this might save us future headaches in case task migrations start happening at bad times
```julia
            if _log_mode() && haslogger()
                tid = Threads.threadid()
                task_log = TaskLog(tj.tag, t₀, t₁, tid, inneighbors_, tj.label)
                push!(_getloginfo().tasklogs[tid], task_log)
                if Threads.threadid() != tid
                    @warning "Thread migration occurred while inserting a TaskLog"
                end
            end
```

What to you think?

**Comment by maltezfaria** [comment]
My future self thanks you very much for this suggestion. I will add it to #61  so that it closes both this issue and #60.
